### PR TITLE
Fix Issue 4587 - Assert exception should not allocate

### PR DESCRIPTION
**Pull request by Darredevil**
Work in Progress 
https://issues.dlang.org/show_bug.cgi?id=4587

### CODE REVIEW

**Comment by Darredevil** [comment]
PS. I noticed there is another PR opened on this topic (https://github.com/dlang/druntime/pull/1710).
Will discuss this with @andralex before moving forward.

File: src/core/exception.d [hunk 1]
```diff
@@ -433,8 +433,15 @@ deprecated void setAssertHandler( AssertHandler h ) @trusted nothrow @nogc
  */
 extern (C) void onAssertError( string file = __FILE__, size_t line = __LINE__ ) nothrow
 {
-    if( _assertHandler is null )
-        throw new AssertError( file, line );
+    import core.stdc.string : memcpy;
+    char[1024] buffer;
```
Comment (lindt) [inline src/core/exception.d]: declare this buffer just within the if

1024 is random and a magic number, you could be concrete with char[__traits(classInstanceSize, AssertError)] buffer;

File: src/core/exception.d [hunk 2]
```diff
@@ -433,8 +433,15 @@ deprecated void setAssertHandler( AssertHandler h ) @trusted nothrow @nogc
...
+    if( _assertHandler is null ) {
+        enum size = __traits(classInstanceSize, AssertError);
+        memcpy(buffer.ptr, typeid(AssertError).initializer().ptr, size);
+        auto obj = cast(AssertError) buffer.ptr;
```
Comment (lindt) [inline src/core/exception.d]: cast is ugly, could emplace be used?

File: src/core/exception.d (hunk 1 above)
Comment (DmitryOlshansky) [inline src/core/exception.d]: Not to mention the alignment issue - I don't think there is guarantee that char[x] array is properly aligned.

Comment (schveiguy) [inline src/core/exception.d]: Not to also mention -- this is a local buffer which you are throwing outside the function. This results in corruption. Make it a static buffer.

File: src/core/exception.d (hunk 2 above)
Comment (schveiguy) [inline src/core/exception.d]: emplace is from Phobos, cannot be used.

**Comment by schveiguy** [comment]
I think we should close this in favor of #1710

**Comment by andralex** [comment]
Yah, let's focus on #1710 and friends.

**Pull request closed without merging**